src/toxcore/messenger/packet/msi.rs

```diff
@@ -0,0 +1,304 @@
+            gen_le_u8!(0x02) >> // MsiError
+            gen_le_u8!(0x01) >> // Size
+            gen_le_u8!(self.0 as u8)
+        )
+    }
+}
+
+impl FromBytes for Capabilities {
+    named!(from_bytes<Capabilities>, do_parse!(
+        tag!("\x03") >>
+        verify!(le_u8, |size| size == 1) >>
+        value: le_u8 >>
+        (Capabilities(value))
+    ));
+}
+
+impl ToBytes for Capabilities {
+    fn to_bytes<'a>(&self, buf: (&'a mut [u8], usize)) -> Result<(&'a mut [u8], usize), GenError> {
+        do_gen!(buf,
+            gen_le_u8!(0x03) >> // Capabilities
+            gen_le_u8!(0x01) >> // Size
+            gen_le_u8!(self.0 as u8)
+        )
+    }
+}
+
+
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+enum MsiSubPacket {
+    Request(Request),
+    MsiError(MsiError),
+    Capabilities(Capabilities),
+}
+
+impl FromBytes for MsiSubPacket {
+    named!(from_bytes<MsiSubPacket>, alt!(
+        map!(Request::from_bytes, MsiSubPacket::Request) |
+        map!(MsiError::from_bytes, MsiSubPacket::MsiError) |
+        map!(Capabilities::from_bytes, MsiSubPacket::Capabilities)
+    ));
+}
+
+impl ToBytes for MsiSubPacket {
+    fn to_bytes<'a>(&self, buf: (&'a mut [u8], usize)) -> Result<(&'a mut [u8], usize), GenError> {
+        match *self {
+            MsiSubPacket::Request(ref p) => p.to_bytes(buf),
+            MsiSubPacket::MsiError(ref p) => p.to_bytes(buf),
+            MsiSubPacket::Capabilities(ref p) => p.to_bytes(buf),
+        }
+    }
+}
+
+impl FromBytes for RequestKind {
+    named!(from_bytes<RequestKind>, switch!(le_u8,
+        1 => value!(RequestKind::Init) |
+        2 => value!(RequestKind::Push) |
+        3 => value!(RequestKind::Pop)
+    ));
+}
+
+impl FromBytes for MsiErrorKind {
+    named!(from_bytes<MsiErrorKind>, switch!(le_u8,
+        0 => value!(MsiErrorKind::MsiNone) |
+        1 => value!(MsiErrorKind::InvalidMessage) |
+        2 => value!(MsiErrorKind::InvalidParam) |
+        3 => value!(MsiErrorKind::InvalidState) |
+        4 => value!(MsiErrorKind::StrayMessage) |
+        5 => value!(MsiErrorKind::System) |
+        6 => value!(MsiErrorKind::Handle) |
+        7 => value!(MsiErrorKind::Undisclosed)
+    ));
+}
+
+/** Msi is a struct that holds info for Media Session Interface.
+
+Sub-packet: kind [1 byte], size [1 byte], value [$size bytes] : but actually size is always 1, so a sub-packet is always 3 bytes long
+
+- kind: one of Request, Capabilities, Error
+- size: the length in byte of value(always 1)
+- value: enum value depending on kind
+
+Payload: |sub_packet| |...{sub-packet}| |0|
+
+Serialized form:
+
+Length    | Content
+--------- | -------
+`1`       | `0x45`
+`0..255`  | payload
+
+Sub-packet serialized form:
+
+Length    | Content
+--------- | -------
+`1`       | kind(1 = Request, 2 = Error, 3 = Capabilities)
+`1`       | size(always 1)
+`1`       | value(it depends on kind)
+
+*/
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct Msi {
+    sub_packets: Vec<MsiSubPacket>,
```

Comment:
I don't like it. How will you interact with it if you need to check that this Msi has capability of `send_video`? `let mut can_send_video = false; for sub in sub_packets { match sub { MsiSubPacket::Capabilities(Capabilities(mask) => if (mask | 32 > 0) { can_send_video = true} } }; `? This looks like OMFG.

Reply:
In c-toxcore, the result of parsing msi packet is stored in other variables.
I can't make decision that we should access Msi struct to get info.

Reply:
In c-toxcore, the processing is done like this manner.
- get one sub-packet from Vec<>
- do the activity defined in sub-packet
- loop

And core maintains other variables of status of msi.

Reply:
Still I don't like it. 

> In c-toxcore, the processing is done like this manner

That's not the answer. They did a lot of architecture mistakes we don't have to copy

Reply:
done